include/commissioner/error.hpp

```diff
@@ -42,36 +43,209 @@ namespace ot {
+     * before finishing the previous one.
+     */
+    kBusy = 9,
+
+    /**
+     * Running out of memory.
+     */
+    kOutOfMemory = 10,
+
+    /**
+     * Read/write to file/socket failed.
+     */
+    kIOError = 11,
+
+    /**
+     * The file/socket is busy and read/write to file/socket will be blocked.
+     */
+    kIOBusy = 12,
+
+    /**
+     * Some entity that we attempted to create (e.g., CoAP resource)
+     * already exists.
+     */
+    kAlreadyExists = 13,
+
+    /**
+     * The operation, transaction or message exchange was aborted.
+     */
+    kAborted = 14,
+
+    /**
+     * The commissioner is not in a valid state that the operation can be processed.
+     */
+
+    kInvalidState = 15,
+
+    /**
+     * The operational has been rejected. For example, petition could be rejected
+     * because of existing active commissioner.
+     */
+    kRejected = 16,
+
+    /**
+     * The error is out of the address space of OT Commissioner.
+     */
+    kUnknown = 17,
 };
 
 /**
- * @brief Convert Error to printable string.
+ * @brief The error of a call in OT Commissioner.
  *
- * @param aError  An error code.
- *
- * @return The printable string of the error.
  */
-const std::string ErrorToString(const Error &aError);
+class OT_COMM_MUST_USE_RESULT Error
+{
+public:
+    /**
+     * The default error is none error.
+     *
+     */
+    Error() {}
+
+    /**
+     * Creates a error with the specified error code and message as a
+     * human-readable string containing more detailed information.
+     *
+     */
+    Error(ErrorCode aErrorCode, const std::string &aErrorMessage);
+
+    // Copy the specified error.
+    Error(const Error &aError);
+    Error &operator=(const Error &aError);
+    Error(Error &&aError) noexcept;
+    Error &operator=(Error &&aError) noexcept;
+
+    /**
+     * Returns true if there is no error.
+     *
+     */
+    bool IsNone() const { return mState == nullptr; }
+
+    ErrorCode GetCode() const { return IsNone() ? ErrorCode::kNone : mState->mCode; }
+
+    const std::string &GetMessage() const { return IsNone() ? EmptyString() : mState->mMessage; }
+
+    /**
+     * Returns a string representation of this error suitable for
+     * printing. Returns the string `"NONE"` for success.
+     *
+     */
+    std::string ToString() const;
+
+    /**
+     * Ignores any errors. This method does nothing except potentially suppress
+     * complaints from any tools that are checking that errors are not dropped on
+     * the floor.
+     *
+     */
+    void IgnoreError() const {}
+
+private:
+    static const std::string &EmptyString();
+
+    struct State
+    {
+        ErrorCode   mCode;
+        std::string mMessage;
+    };
+
+    // `None` error has a `NULL` mState. Otherwise, `mState` points to
+    // a `State` structure containing the error code and message(s).
+    // This is for performance, because a `None` error is the usual case.
+    std::unique_ptr<State> mState;
```

Comment:
What is the reason behind the extra indirection and unique pointer use here?

Wouldn't using an `ErrorCode mCode` and `std::string mMessage` in `Error` be simpler and and avoid the dynamic allocation?

Reply:
In most situations, we will success and error will be `None`.  `std::string mMessage` typically occupies 32 bytes on 64 bits platform and the constructor or copy assignment will be called when the error is initialized or copied. The `unique_ptr` here is to make the error class smaller and avoid unnecessary copy.

FYI, our implementation here is inspired by [abseil/status.h](https://github.com/abseil/abseil-cpp/blob/master/absl/status/status.h) which has a similar imeplementation.

Reply:
I guess this model is intended for use in code running on POSIX MCU/machines (i.e. not as resource constrained as for example OT core which can run on rather resource limited MCUs). So having an extra 32 bytes of memory per `Error` may be fine (particularly since `Error` instances are mostly temp/stack variables). 

On the other hand, I would expect the `std::string` copy constructor implementation to be efficient (specially for empty strings) and the advantage of using `ErrorCode mCode` and `std::string mMessage` is that the code would be simple and  avoids dynamic allocation and use of `unique_ptr` (which add their own overhead in terms of speed).

Reply:
@abtink Agree that size and performance are not that important in context of an external Thread Commissioner. I removed the `unique_ptr`, please review.

Reply:
But it-commissioner is indeed supposed to run on posix, so unique_ptr should be proper.